### ISSUE TITLE
Panic on overflowing index with Rust <1.45

### ISSUE THREAD

**Issue by adamsky** [adamsky/globe#8]
As reported on reddit by [dd-pardal](https://www.reddit.com/r/rust/comments/jj0pde/heres_a_fun_little_crate_i_made_that_generates_an/gaack5n/?utm_source=reddit&utm_medium=web2x&context=3).

>That's great. I don't know if it's just me, but when I pan the globe up, the code panics:
>It seems like it's trying to get the character for an absurdly high luminance.
>I'm using 64-bit Windows.

```
thread 'main' panicked at 'index out of bounds: the len is 18 but the index is 18446744073709551615', globe\src\lib.rs:141:43
```

**Comment by James-Livesey** [comment]
Exact same error for me when I attempt to rotate the globe with either my mouse or keyboard (the rotation works for small amounts, but the error throws when I rotate it quickly). To capture the error trace, I ran:

```bash
$ RUST_BACKTRACE=1 globe -i 2> globe.log
```

The contents of `globe.log` is:

```
thread 'main' panicked at 'index out of bounds: the len is 18 but the index is 18446744073709551615', /home/james/.cargo/registry/src/github.com-1ecc6299db9ec823/globe-0.1.1/src/lib.rs:137:43
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /usr/src/rustc-1.41.1/vendor/backtrace/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /usr/src/rustc-1.41.1/vendor/backtrace/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:84
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:61
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1025
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1426
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:65
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:50
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:193
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:210
  10: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:471
  11: rust_begin_unwind
             at src/libstd/panicking.rs:375
  12: core::panicking::panic_fmt
             at src/libcore/panicking.rs:84
  13: core::panicking::panic_bounds_check
             at src/libcore/panicking.rs:62
  14: globe::Globe::render_on
  15: globe::main
  16: std::rt::lang_start::{{closure}}
  17: std::rt::lang_start_internal::{{closure}}
             at src/libstd/rt.rs:52
  18: std::panicking::try::do_call
             at src/libstd/panicking.rs:292
  19: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:78
  20: std::panicking::try
             at src/libstd/panicking.rs:270
  21: std::panic::catch_unwind
             at src/libstd/panic.rs:394
  22: std::rt::lang_start_internal
             at src/libstd/rt.rs:51
  23: main
  24: __libc_start_main
  25: _start
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

Awesome project though!

---
Edited to add: Debian 10 Linux, lxterminal

**Comment by adamsky** [comment]
Thanks for your input! There is a makeshift fix available on the [pan-up-panic](https://github.com/adamsky/globe/tree/pan-up-panic) branch, but I couldn't get the last person who got the error to report back on whether it fixes the problem.

**Comment by jerrytingley** [comment]
Found this ticket due to having the same issue. Just compiled the pan-up-panic branch and is working for me on Ubuntu 20.04. I can pan up with no panic or error messages.

**Comment by and-reas-se** [comment]
Worth noting that the wrong index is equal to 2^64 - 1, the highest number you can store in an usize on common architectures.

**Comment by adamsky** [comment]
@jerrytingley There's a more elegant potential fix at [fix-luminance](https://github.com/adamsky/globe/tree/fix-luminance) branch now, could you please confirm it fixes the issue?

**Comment by jamesblau** [comment]
For me, pan-up-panic fixes it, but not fix-luminance, with kitty terminal (TERM=xterm-kitty) or gnome-terminal (TERM=xterm-256color).

With suckless simple terminal (TERM=st-256color), fix-luminance and master don't even render before giving the same error, but pan-up-panic does still work.

Ubuntu 20.04.1 LTS, cargo 1.44.1, rustup 1.22.0 if that matters.

**Comment by adamsky** [comment]
I'll need to find a machine where I can replicate this, there's an overflow happening somewhere but I'm not sure where exactly. For now I'll add the temporary fix so that it clamps the index instead of outright panicking.

**Comment by adamsky** [comment]
Oh, one last try. @jamesblau could you please test the `fix-luminance` branch once more and let me know if there is any difference?

**Comment by and-reas-se** [comment]
Some wild speculation:

I think you have a mistake in the math somewhere, a division by zero or something along those lines, that is producing a NaN. Then you're casting that to an integer, and integer values do not have NaNs. Before rust 1.45 this was undefined behavior, so getting a weird number out of that cast is reasonable. In rust 1.45 and later they decided to get rid of this undefined behavior so they defined casting NaN to integer would produce a 0 (discussion in release notes: https://blog.rust-lang.org/2020/07/16/Rust-1.45.0.html). That's why you can't reproduce, you're using a version of rust since after the change.

**Comment by adamsky** [comment]
@AndreasTPC That sounds about right, thanks! I'll need to follow those changelogs more closely.

**Comment by adamsky** [comment]
Alright so https://github.com/adamsky/globe/commit/ac55296a0d5c12c8e309cc9cc56f274bb9ef04be fixes the panic issue for Rust versions <1.45. Math remains unsound, but that's an issue for another time.
